### PR TITLE
Feature/#9

### PR DESCRIPTION
**Pull request by hamzakhawaja27**
Code commented
Wiki edited
Up for review.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -15,3 +15,5 @@ dist/*
 #osx operating system files
 *.[Dd][Ss]_[Ss]tore*
 
+#pycharm config files
+.idea/*
```
Comment (jamesfoote) [inline .gitignore]: Don't need these here, they aren't necessary for just using the repo.

File: src/ucapi/ecs_client.py [hunk 2]
```diff
@@ -280,11 +283,78 @@ def _delete_launch_configuration(self, group_name):
...
+        :param: cluster_name= None
+
+        Objective: run the task,only.
+        :param: task_definition= Path for taskdefinition.json of a task to be run or taskARN of the task to be deployed.
+        :param: cluster_name= name of the cluster in which the task is to be deployed
+
+        '''
+
+        if isinstance(task_definition, dict) and cluster_name is not None:
+            try:
```
Comment (jamesfoote) [inline src/ucapi/ecs_client.py]: Probably shouldn't be try-except here. Better idea to do something like:
`if 'containerDefinitions' in task_definition:

...

else:

...`

File: src/ucapi/ecs_client.py [hunk 3]
```diff
@@ -280,11 +283,78 @@ def _delete_launch_configuration(self, group_name):
...
+                # if task definition is a container definition then reg and run the task
+                container_definitions=task_definition
+                family_name = container_definitions['name']
+                response1 = self._aws_ecs_client.register_task_definition(containerDefinitions=[container_definitions],
+                                                                          family=family_name)
+                task_arn = response1["taskDefinition"]["taskDefinitionArn"]
+                response2 = self._aws_ecs_client.run_task(taskDefinition=task_arn, cluster=cluster_name)
+                return response2
+
+        elif (isinstance(task_definition, dict) or isinstance(task_definition, list)) and cluster_name is None:
```
Comment (jamesfoote) [inline src/ucapi/ecs_client.py]: What situation would the task_definition be a list?

File: src/ucapi/ecs_client.py [hunk 4]
```diff
@@ -280,11 +283,78 @@ def _delete_launch_configuration(self, group_name):
...
+            elif isinstance(task_definition, list):
+                container_definitions=task_definition
+                family_name=container_definitions[0]["name"]
+                response = self._aws_ecs_client.register_task_definition(containerDefinitions=container_definitions,
+                                                                         family=family_name)
+                return response
+
+        elif isinstance(task_definition, str) and cluster_name is not None:
+            # taskDefinition=TaskDefinitionPath given, we run the task.
+            try:
```
Comment (jamesfoote) [inline src/ucapi/ecs_client.py]: Try-except best used for error catching, try find some other logic to determine whether or not the str is a file.

File: src/ucapi/ecs_client.py [hunk 5]
```diff
@@ -280,11 +283,78 @@ def _delete_launch_configuration(self, group_name):
...
+                return response
+            except:
+                try:
+                    task_arn=task_definition
+                    response = self._aws_ecs_client.run_task(taskDefinition=task_arn, cluster=cluster_name)
+                    return response
+                except:
+                    raise InvalidArgumentException("Wrong TaskARN")
+        else:
+            raise InvalidArgumentException("wrong input parameters")
```
Comment (jamesfoote) [inline src/ucapi/ecs_client.py]: Change to "Wrong input parameters."

File: src/ucapi/ecs_client.py (hunk 3 above)
Comment (hamzakhawaja27) [inline src/ucapi/ecs_client.py]: Task definition could be a list if the client need to deploy several container with one command.

File: src/ucapi/ecs_client.py (hunk 2 above)
Comment (hamzakhawaja27) [inline src/ucapi/ecs_client.py]: Made the change.

File: .gitignore (hunk 1 above)
Comment (hamzakhawaja27) [inline .gitignore]: made the changes

File: src/ucapi/ecs_client.py (hunk 4 above)
Comment (hamzakhawaja27) [inline src/ucapi/ecs_client.py]: changed to 
if os.path.exists(task_definition):

File: src/ucapi/ecs_client.py (hunk 5 above)
Comment (hamzakhawaja27) [inline src/ucapi/ecs_client.py]: Done.

File: src/ucapi/ecs_client.py (hunk 3 above)
Comment (jamesfoote) [inline src/ucapi/ecs_client.py]: It looks like you are just accessing the head of the list, then ignoring the rest of it.

File: src/ucapi/ecs_client.py [hunk 6]
```diff
@@ -280,11 +284,78 @@ def _delete_launch_configuration(self, group_name):
...
+                response1 = self._aws_ecs_client.register_task_definition(containerDefinitions=container_definitions,
+                                                                          family=family_name)
+                task_arn = task_definition['taskDefinitionArn']
+                response2 = self._aws_ecs_client.run_task( taskDefinition=task_arn, cluster=cluster_name)
+                return response2
+            else:
+                # if task definition is a container definition then reg and run the task
+                container_definitions=task_definition
+                family_name = container_definitions['name']
+                response1 = self._aws_ecs_client.register_task_definition(containerDefinitions=[container_definitions],
```
Comment (jamesfoote) [inline src/ucapi/ecs_client.py]: Remove the response1, response2 etc notation throughout, just write over the previous response value unless you need it for something.

File: .gitignore (hunk 1 above)
Comment (jamesfoote) [inline .gitignore]: Did you push them? I can still see the lines :P

File: src/ucapi/ecs_client.py (hunk 6 above)
Comment (hamzakhawaja27) [inline src/ucapi/ecs_client.py]: Done.

File: src/ucapi/ecs_client.py (hunk 3 above)
Comment (hamzakhawaja27) [inline src/ucapi/ecs_client.py]: response = self._aws_ecs_client.register_task_definition(containerDefinitions=[container_definitions],family=family_name).

Here the container definition has to be a list. If I get a dict, I would insert it in a list and then pass it on to this function. This function registers all the containers which are defined in the list.

Comment (jamesfoote) [inline src/ucapi/ecs_client.py]: Okay easy, my bad

File: .gitignore (hunk 1 above)
Comment (hamzakhawaja27) [inline .gitignore]: Yeah you were right. Can you check again if the changes are made.

**Pull request merged**